// leasot/dist/cli/cli.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFiletype = void 0;
const get_stdin_1 = __importDefault(require("get-stdin"));
const globby_1 = __importDefault(require("globby"));
const log_symbols_1 = __importDefault(require("log-symbols"));
const __1 = require("..");
const path_1 = require("path");
const async_1 = require("async");
const fs_1 = require("fs");
const common_1 = require("./common");
const DEFAULT_EXTENSION = '.js';
const CONCURRENCY_LIMIT = 50;
/**
 * @hidden
 */
const getFiletype = (filetype, filename) => {
    if (filetype) {
        return filetype;
    }
    if (filename && path_1.extname(filename)) {
        return path_1.extname(filename);
    }
    return DEFAULT_EXTENSION;
};
exports.getFiletype = getFiletype;
const parseContentSync = (content, options, filename) => {
    const extension = exports.getFiletype(options.filetype, filename);
    __1.associateExtWithParser(options.associateParser);
    if (!__1.isExtensionSupported(extension)) {
        if (options.skipUnsupported) {
            return [];
        }
        console.log(log_symbols_1.default.error, `Filetype ${extension} is unsupported.`);
        process.exit(1);
    }
    const config = {
        customTags: options.tags,
        extension: extension,
        filename: filename,
        withInlineFiles: options.inlineFiles,
    };
    return __1.parse(content, config);
};
const parseAndReportFiles = (fileGlobs, options) => {
    // Get all files and their resolved globs
    const files = globby_1.default.sync(fileGlobs, {
        ignore: options.ignore || [],
    });
    if (!files || !files.length) {
        console.log(log_symbols_1.default.warning, 'No files found for parsing');
        process.exit(1);
    }
    // Parallel read all of the given files
    async_1.mapLimit(files, CONCURRENCY_LIMIT, (file, cb) => fs_1.readFile(path_1.resolve(process.cwd(), file), 'utf8', cb), (err, results) => {
        if (err) {
            console.log(err);
            process.exit(1);
        }
        const todos = results
            .map(function (content, index) {
            return parseContentSync(content, options, files[index]);
        })
            // filter files without any parsed content
            .filter(item => item && item.length > 0)
            .reduce((items, item) => items.concat(item), []);
        common_1.outputTodos(todos, options);
    });
};
const run = (program) => {
    const options = program.opts();
    if (program.args && program.args.length > 0) {
        return parseAndReportFiles(program.args, options);
    }
    if (process.stdin.isTTY) {
        return program.help();
    }
    // data is coming from a pipe
    get_stdin_1.default()
        .then(function (content) {
        const todos = parseContentSync(content, options);
        common_1.outputTodos(todos, options);
    })
        .catch(function (e) {
        console.error(e);
        process.exit(1);
    });
};
exports.default = run;
