// leasot/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.prepareComment = exports.getRegex = exports.report = exports.parse = exports.isExtensionSupported = exports.associateExtWithParser = void 0;
var parsers_1 = require("./lib/parsers");
Object.defineProperty(exports, "associateExtWithParser", { enumerable: true, get: function () { return parsers_1.associateExtWithParser; } });
Object.defineProperty(exports, "isExtensionSupported", { enumerable: true, get: function () { return parsers_1.isExtensionSupported; } });
Object.defineProperty(exports, "parse", { enumerable: true, get: function () { return parsers_1.parse; } });
var reporters_1 = require("./lib/reporters");
Object.defineProperty(exports, "report", { enumerable: true, get: function () { return reporters_1.report; } });
var utils_1 = require("./lib/utils");
Object.defineProperty(exports, "getRegex", { enumerable: true, get: function () { return utils_1.getRegex; } });
Object.defineProperty(exports, "prepareComment", { enumerable: true, get: function () { return utils_1.prepareComment; } });
